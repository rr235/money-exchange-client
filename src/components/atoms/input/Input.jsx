import React, { useState, FormEvent } from 'react';
import classNames from 'classnames';
import styles from './input.styles.scss';

const Input = ({ label, id, onChange, className }) => {
  const [input, setInput] = useState('');

  const onChangeHandler = (e) => {
    setInput(e.currentTarget.value);
    if (onChange) {
      onChange(e);
    }
  };

  return (
    <>
      <label htmlFor={id} className={classNames(styles.label, className)}>
        {label}
      </label>
      <input
        type="number"
        id={id}
        className={styles.input}
        value={input}
        onChange={onChangeHandler}
      />
    </>
  );
};

export default Input;
