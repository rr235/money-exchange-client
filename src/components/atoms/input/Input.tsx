import React, { useState, FormEvent } from 'react';
import classNames from 'classnames';
import styles from './input.styles.scss';

type Props = {
  id: string;
  label: string;
  className?: string;
  onChange?(e: FormEvent<HTMLInputElement>): () => void;
};

const Input = ({ label, id, onChange, className }: Props) => {
  const [input, setInput] = useState('');

  const onChangeHandler = (e: FormEvent<HTMLInputElement>): void => {
    setInput(e.currentTarget.value);
    if (onChange) {
      onChange(e);
    }
  };

  return (
    <>
      <label htmlFor={id} className={classNames(styles.label, className)}>
        {label}
      </label>
      <input
        type="number"
        id={id}
        className={styles.input}
        value={input}
        onChange={onChangeHandler}
      />
    </>
  );
};

export default Input;
