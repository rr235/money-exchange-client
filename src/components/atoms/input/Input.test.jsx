import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Input from './Input';

describe('Input ', () => {
  let component;
  beforeAll(() => {
    component = shallow(<Input label="label" id="id" type="number" />);
  });

  it('should render correctly (type="number")', () => {
    const tree = renderer
      .create(<Input label="label" id="id" type="number" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly (type="text")', () => {
    const tree = renderer
      .create(<Input label="label" id="id" type="text" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should contain label', () => {
    expect(component.find('label')).toHaveLength(1);
  });

  it('should contain input field', () => {
    expect(component.find('input[type="number"]')).toHaveLength(1);
  });

  it('should show correct label', () => {
    expect(component.find('label').text()).toBe('label');
  });
});

describe('Input (number) - Functionality', () => {
  let component;
  beforeAll(() => {
    act(() => {
      component = shallow(<Input label="label" id="id" type="number" />);
    });
  });

  it('should take correct input', () => {
    let input = component.find('input[type="number"]');

    // simulate on change with valid input
    act(() => {
      input.props().onChange({
        currentTarget: {
          value: 100,
        },
      });
    });

    // update state and get latest ref to input
    component.update();
    input = component.find('input[type="number"]');

    expect(input.prop('value')).toBe(100);
  });

  it('should show previous input when input is invalid', () => {
    let input = component.find('input[type="number"]');

    // simulate on change with VALID input
    act(() => {
      input.props().onChange({
        currentTarget: {
          value: 100,
        },
      });
    });

    // update state and get latest ref to input
    component.update();
    input = component.find('input[type="number"]');
    expect(input.prop('value')).toBe(100);

    // simulate on change with INVALID input
    act(() => {
      input.props().onChange({
        currentTarget: {
          value: '200,222',
        },
      });
    });

    // update state and get latest ref to input
    component.update();
    input = component.find('input[type="number"]');

    expect(input.prop('value')).toBe(100);
  });
});
