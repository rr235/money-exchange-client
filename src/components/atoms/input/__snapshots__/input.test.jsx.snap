// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input  should render correctly (type="number") 1`] = `
Array [
  <label
    className="label"
    htmlFor="id"
  >
    label
  </label>,
  <input
    className="input"
    id="id"
    onChange={[Function]}
    type="number"
    value=""
  />,
]
`;

exports[`Input  should render correctly (type="text") 1`] = `
Array [
  <label
    className="label"
    htmlFor="id"
  >
    label
  </label>,
  <input
    className="input"
    id="id"
    onChange={[Function]}
    type="text"
    value=""
  />,
]
`;
